provider/awssd/aws_sd.go

```diff
@@ -193,6 +211,12 @@ func (p *AWSSDProvider) instancesToEndpoint(ns *sd.NamespaceSummary, srv *sd.Ser
 			newEndpoint.RecordType = endpoint.RecordTypeCNAME
 			newEndpoint.Targets = append(newEndpoint.Targets, aws.StringValue(inst.Attributes[sdInstanceAttrAlias]))
 
+			// SRV
+		} else if inst.Attributes[sdInstanceAttrOriginalSrv] != nil && aws.StringValue(srv.DnsConfig.DnsRecords[0].Type) == sd.RecordTypeSrv {
+			newEndpoint.RecordType = endpoint.RecordTypeSRV
+			newEndpoint.Targets = append(newEndpoint.Targets, p.srvIPTargetCombine(
+				aws.StringValue(inst.Attributes[sdInstanceAttrIPV4]), aws.StringValue(inst.Attributes[sdInstanceAttrOriginalSrv])))
+
```

Comment:
I wonder if it would better to convert each Cloud Map SRV instance to two endpoints right away (SRV+A). And then remove duplicate SRV endpoints (since all SRV endpoints sourced from a single instance are going to be identical). Maybe the complicated logic in `srvEndpointsIpToHost` could be avoided.

Reply:
Thanks a lot for your comment. In my proposal, `srvEndpointsIpToHost` is needed to handle cases when Cloud Map SRV instances are not aligned to A instances. Let's take the example below:

1. At the beginning, cluster has 1 node: 10.0.0.1. Service nginx uses nodePort 30080. CloudMap is up-to-date on both A and SRV instances, so `Records` returns the following endpoints: `nginx.abc.local 300 IN A  10.0.0.1` and `_http._tcp.nginx.abc.local 300 IN SRV  0 50 30080 nginx.abc.local`. The endpoints reported are in line with expected, so no change is proposed by planner.
2. A new node 10.0.0.2 is added. The planner generates a change to update the A endpoint to `nginx.abc.local 300 IN A  10.0.0.1;10.0.0.2`. But no change plan is generated for the SRV record.
3. Planned change for A record is implemented in Cloud Map by `ApplyChanges`. Now Cloud Map SRV instances are not aligned with Cloud Map A instances (SRV has 1 instance and A has 2). Here `srvEndpointsIpToHost` comes into play as it catches the misalignment and enables `Records` to report the SRV endpoint as `_http._tcp.nginx.abc.local 300 IN SRV  10.0.0.1_0_50_30080_nginx.abc.local`. The SRV endpoint reported is not in line with expected (`_http._tcp.nginx.abc.local 300 IN SRV  0 50 30080 nginx.abc.local`), so a change is proposed by planner.
4. Planned change for SRV record is implemented in Cloud Map by `ApplyChanges`. Now Cloud Map has two SRV instances and is aligned with A record, so it is reported by `Records` as `_http._tcp.nginx.abc.local 300 IN SRV  0 50 30080 nginx.abc.local`. The SRV endpoint reported is in line with expected, so no change is proposed by planner.

> I wonder if it would better to convert each Cloud Map SRV instance to two endpoints right away (SRV+A). And then remove duplicate SRV endpoints (since all SRV endpoints sourced from a single instance are going to be identical). Maybe the complicated logic in `srvEndpointsIpToHost` could be avoided.

Now, if I understand your comment correctly, I think it would fail on point 3 to catch misaligned SRV instances. As a result, the planner would not produce the expected update for the SRV record. Looking forward for any feedback.